### PR TITLE
add Hierarchical Clustering & some docstring fixes

### PR DESCRIPTION
**Pull request by jbrea**
@ablaom `predict(mach, Xnew)` doesn't make any sense for hierarchical clustering, but `predict(mach)` does. Do you think this is still useful like this?

(ps. Sorry for mixing in docstring fixes...)

### CODE REVIEW

**Comment by ablaom** [comment]
@jbrea Thanks indeed for this contribution.

On a first glance this implementation makes sense, but after some reflection, certain aspects bother me. For example, the possibility that  `predict(mach)` can give different answers without `fit!` being called in between sounds dangerous. (This should probably be ruled out in the API.) Although this is a weak argument, another objection is that type ambiguity prevents us from having operations with zero data arguments - you would have to call `predict(mach, nothing)` or something. 

Instead, I would be inclined to conceptualise this model as a [`Static` transformer](https://alan-turing-institute.github.io/MLJ.jl/dev/transformers/#Static-transformers). After all, you are not "learning" in the sense of "learning to generalise to unseen examples". Implemented this way, there is no `fit` to overload, just a `transform` method with the "training" data as input. There remains the question of what the output should be. Some options I can think of:

1. Just return the full  `Hclust` object. Pros: flexibility to determine the right cut-off h/k in a later process. Con: To use the object, the user needs to be familiar with its interface
2. Return the cluster labels, as you do now, according to the h/k values in the struct. Pros: Simple. Interpretation of output is clear. Cons: Need to re-compute (recall `transform`) for every new h/k.
3. Return a closure that maps `(k, h)` or `(k, )` or `(h, )` to the labels for the cutoff defined by `k, h`. Pros: simple, flexible. No need for user to know the interface for `Hclust`. Cons: It is unusual to return callable objects. However, a nice `show` for the closure could mitigate the mystery. We did this for DecisionTrees [here](https://github.com/JuliaAI/MLJDecisionTreeInterface.jl/blob/66296193de2b22c162f25192a3bc2a42facccf3c/src/MLJDecisionTreeInterface.jl#L16) where the closure is for pretty printing of the tree and the argument is the tree depth. (Aside, there is already a scitype for callable objects, `CallableReturning{S}`, but I need to move it from MLJBase into ScientificTypesBase, which I'm happy to do.) **edit** I guess we would need to write a plot recipe to overload `plot` for our callable object. 

My vote is for 3. What do you think? And do you have objectionis to implementing as `Static`?

**Comment by jbrea** [comment]
@ablaom Thanks for your feedback and suggestions! I like version 3 and implemented it.
I think the plot recipe is not urgent with `plot(hc.dendrogram)` (see docstring of `HierarchicalClustering`).
The only thing I dislike about this approach is that it is not a model and therefore not listed in `models()` and e.g. `info("HierarchicalClustering")` errs.

**Comment by ablaom** [comment]
@jbrea Thanks for that. Unfortunately, since making the suggestion I'm rethinking my `Static` suggestion. After discussions at the other open PR on DBSCAN, I'm not sure this is best after all. And I see that all the sklearn static clusterers that we wrap are implemented as `Unsupervised` transformers without `transform` or `predict` methods. 

Can we sit on this until the API is [firmed up](https://github.com/alan-turing-institute/MLJ.jl/issues/852)?

**Comment by ablaom** [comment]
@jbrea Thanks for your patience. We may want to revisit this after https://github.com/JuliaAI/MLJBase.jl/pull/806

**Comment by jbrea** [comment]
I'll have a look at this in early September, if you are not in a hurry.

**Comment by jbrea** [comment]
Hi @ablaom 

I am struggling a bit with squeezing hierarchical clustering into the current design.
In hierarchical clustering the "heavy-lifting" is done when computing the dendrogram.
Once one has the dendrogram, it is cheap to make predictions by cutting the dendrogram at the desired height.
The current implementation uses a `_cache` field that holds the dendrogram and a hash of the data used to compute the dendrogram in order to trigger recomputation of the dendrogram, if the data changes significantly.
Do you see a better approach?

File: src/MLJClusteringInterface.jl [hunk 1]
```diff
@@ -31,13 +31,14 @@ const PKG = "MLJClusteringInterface"
...
     metric::SemiMetric = SqEuclidean()
+    init = :kmpp
 end
 
 function MMI.fit(model::KMeans, verbosity::Int, X)
     # NOTE: using transpose here to get a LinearAlgebra.Transpose object
     # which Kmeans can handle.
     Xarray = transpose(MMI.matrix(X))
-    result = Cl.kmeans(Xarray, model.k; distance=model.metric)
+    result = Cl.kmeans(Xarray, model.k; distance=model.metric, init = model.init)
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]: As orthogonal to hierarchical clustering, please move this change to a separate PR an include an update to the (recently expanded) doc-string.

**Comment by ablaom** [comment]
@jbrea Thanks again for looking into this. I agree, this is just not easy to fit into the current design (or any general ML API, really). 
I really like your idea, and it's probably the best solution from the point of view of user-friendliness. 

However, I'm pretty sure storing state in `model` is going to lead to unexpected behaviour down the line. That is, I think `predict` needs to re-compute from scratch everytime it is called. 

I think the best we can do is include your earlier closure in the `report` (in addition to the raw `Clustering.Hclust` object). The `predict` call should return the assignments corresponding to the (`k`, `c`) parameters at time of call. 

What do you say?

**Comment by jbrea** [comment]
Thanks, @ablaom. This sounds good to me. I pushed the new version.

File: src/MLJClusteringInterface.jl [hunk 2]
```diff
@@ -177,27 +177,36 @@ MMI.reporting_operations(::Type{<:DBSCAN}) = (:predict,)
...
+"""
+    (cutter::DendrogramCutter)(; h = nothing, k = 3)
+
+Cuts the dendrogram at height `h` or, if `height == nothing`, such that `k` clusters are obtained.
+"""
+function (cutter::DendrogramCutter)(; h = nothing, k = 3)
+    MMI.categorical(Cl.cutree(cutter.dendrogram, k = k, h = h))
+end
+function Base.show(io::IO, ::DendrogramCutter)
+    println(io, "Dendrogram Cutter.")
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]: Shouldn't `println`  be just `print` here?
```suggestion
    print(io, "Dendrogram Cutter.")
```

File: src/MLJClusteringInterface.jl [hunk 3]
```diff
@@ -238,6 +247,13 @@ metadata_model(
     path = "$(PKG).DBSCAN"
 )
 
+metadata_model(
+    HierarchicalClustering,
+    human_name = "Hierarchical clusterer",
+    input = MMI.Table(Continuous),
+    output = MMI.Table(Continuous),
+    path = "$(PKG).HierarchicalClustering"
+)
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]:
```suggestion
metadata_model(
    HierarchicalClustering,
    human_name = "hierarchical clusterer",
    input = MMI.Table(Continuous),
    target = MMI.AbstractVector{<:Multiclass}
    path = "$(PKG).HierarchicalClustering"
)
```

File: src/MLJClusteringInterface.jl [hunk 4]
```diff
@@ -510,4 +526,81 @@ scatter(points, color=colors)
...
+
+- `k = 3`: number of clusters; this is ignored, if `h != nothing`.
+
+
+# Operations
+
+- `predict(mach, X)`: return cluster label assignments, as an unordered
+  `CategoricalVector`. Here `X` is any table of input features (eg, a `DataFrame`) whose
+  columns are of scitype `Continuous`; check column scitypes with `schema(X)`. Note that
+  points of type `noise` will always get a label of `0`.
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]: I guess this is left-over from DBSCAN.
```suggestion
  columns are of scitype `Continuous`; check column scitypes with `schema(X)`. 
```

File: src/MLJClusteringInterface.jl [hunk 5]
```diff
@@ -510,4 +526,81 @@ scatter(points, color=colors)
...
+
+
+# Report
+
+After calling `predict(mach)`, the fields of `report(mach)`  are:
+
+- `dendrogram`: the dendrogram that was computed when calling `predict`.
+
+- `cutter`: a dendrogram cutter that can be called with a height `h` or a number of clusters `k`, to obtain a new assignment of the data points to clusters.
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]:
```suggestion
- `cutter`: a dendrogram cutter that can be called with a height `h` or a number of clusters `k`, to obtain a new assignment of the data points to clusters (see example below).
```

File: src/MLJClusteringInterface.jl [hunk 6]
```diff
@@ -510,4 +526,81 @@ scatter(points, color=colors)
...
+using StatsPlots
+plot(report(mach).dendrogram)
+
+# make new predictions by cutting the dendrogram at another height
+report(mach).cutter(h = 2.5)
+```
+
+"""
+HierarchicalClustering
+
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl:601]: Beautiful doc-string.

File: src/MLJClusteringInterface.jl [hunk 7]
```diff
@@ -477,4 +526,81 @@ scatter(points, color=colors)
...
+# Examples
+
+```
+using MLJ
+
+X, labels  = make_moons(400, noise=0.09, rng=1) # synthetic data with 2 clusters; X
+y = map(labels) do label
+    label == 0 ? "cookie" : "monster"
+end;
+y = coerce(y, Multiclass);
```
Comment (ablaom) [inline src/MLJClusteringInterface.jl]: Unless you want to add some comparison to "true" labels, we can drop this as redundant, yes?
```suggestion
```

**Review by ablaom — COMMENTED** [review]
Nice implementation. Perhaps we should add some tests?

At the least, we should add `HierarchicalClusterer` to the generic interface tests at the end of `runtests.jl`

- [x] https://github.com/jbrea/MLJClusteringInterface.jl/blob/61a249934465878fb6d5b10ea24309f1b5aca1bd/test/runtests.jl#L99

We also need to update this line to include the new model:

- [x] https://github.com/JuliaAI/MLJClusteringInterface.jl/blob/6f60a3596e210f1079f7fccc0b8fb81d235ffe5e/src/MLJClusteringInterface.jl#L175

File: src/MLJClusteringInterface.jl (hunk 6 above)
Comment (jbrea) [reply src/MLJClusteringInterface.jl:601]: Thanks :) I had a nice template to copy from :)

File: src/MLJClusteringInterface.jl (hunk 3 above)
Comment (jbrea) [reply src/MLJClusteringInterface.jl]: Actually, does the target make any sense here? DBSCAN also doesn't have it. I removed it for now.

**Comment by jbrea** [comment]
Thanks, @ablaom. Tests are added now. I think this PR should be ready now.

File: src/MLJClusteringInterface.jl (hunk 3 above)
Comment (ablaom) [reply src/MLJClusteringInterface.jl]: You're probably correct: The current spec only talks about "target" in the context of supervised learning. 

Nowadays, I think of a target as anything that gets paired with a like-variable (or a proxy ,such as a probability distribution) to get some kind of numerical "score". In clustering, such a score is the Rand index. So, models with a target, including clusterers, can be "evaluated", even though if they are not supervised. I am working this view into a proposed refactoring of MLJ's foundations (which get's rid of the model hierarchy, among other things).

Thanks for pointing out the inconsistency with DBSCAN. I shall look into this and open a separate issue if necessary. So what you have here is fine for now.

**Review by ablaom — APPROVED** [review]
Good to go. @jbrea Thanks for this valuable contribution, your patience, and careful design considerations.
